File: packages/mpc-wallet-sdk/src/policy.ts
/**
 * Policy engine for transaction validation
 */

import type { ChainType, TransactionRequest } from './types';

/**
 * Policy decision result
 */
export interface PolicyDecision {
  /** Whether transaction is approved */
  approved: boolean;
  /** Whether additional approval is required */
  requiresAdditionalApproval: boolean;
  /** Reason for rejection (if any) */
  reason?: string;
}

/**
 * Spending limits configuration
 */
export interface SpendingLimits {
  /** Per-transaction limit (in wei) */
  perTransaction?: bigint;
  /** Daily limit (in wei) */
  daily?: bigint;
  /** Weekly limit (in wei) */
  weekly?: bigint;
  /** Currency symbol */
  currency: string;
}

/**
 * Time window restriction
 */
export interface TimeBounds {
  /** Start hour (0-23, UTC) */
  startHour: number;
  /** End hour (0-23, UTC) */
  endHour: number;
  /** Allowed days (0=Sunday, 6=Saturday) */
  allowedDays: number[];
}

/**
 * Contract interaction restriction
 */
export interface ContractRestriction {
  /** Allowed contract addresses (empty = all allowed) */
  allowedContracts: Set<string>;
  /** Allowed function selectors (4-byte hex) */
  allowedSelectors: Set<string>;
  /** Blocked function selectors */
  blockedSelectors: Set<string>;
}

/**
 * Policy configuration
 */
export class PolicyConfig {
  spendingLimits: Map<ChainType, SpendingLimits> = new Map();
  whitelist: Set<string> | null = null;
  blacklist: Set<string> = new Set();
  timeBounds: TimeBounds | null = null;
  contractRestrictions: ContractRestriction | null = null;
  additionalApprovalThreshold: bigint | null = null;
  maxPendingRequests: number = 10;
  enabled: boolean = true;

  /**
   * Create a new policy config
   */
  static create(): PolicyConfig {
    return new PolicyConfig();
  }

  /**
   * Create a disabled policy (all transactions allowed)
   */
  static disabled(): PolicyConfig {
    const config = new PolicyConfig();
    config.enabled = false;
    return config;
  }

  /**
   * Set per-transaction spending limit
   */
  withPerTxLimit(amount: bigint, currency = 'ETH'): PolicyConfig {
    const limits = this.spendingLimits.get(0) ?? { currency };
    limits.perTransaction = amount;
    limits.currency = currency;
    this.spendingLimits.set(0, limits);
    return this;
  }

  /**
   * Set daily spending limit
   */
  withDailyLimit(amount: bigint, currency = 'ETH'): PolicyConfig {
    const limits = this.spendingLimits.get(0) ?? { currency };
    limits.daily = amount;
    limits.currency = currency;
    this.spendingLimits.set(0, limits);
    return this;
  }

  /**
   * Set weekly spending limit
   */
  withWeeklyLimit(amount: bigint, currency = 'ETH'): PolicyConfig {
    const limits = this.spendingLimits.get(0) ?? { currency };
    limits.weekly = amount;
    limits.currency = currency;
    this.spendingLimits.set(0, limits);
    return this;
  }

  /**
   * Set address whitelist
   */
  withWhitelist(addresses: string[]): PolicyConfig {
    this.whitelist = new Set(addresses.map((a) => a.toLowerCase()));
    return this;
  }

  /**
   * Set address blacklist
   */
  withBlacklist(addresses: string[]): PolicyConfig {
    this.blacklist = new Set(addresses.map((a) => a.toLowerCase()));
    return this;
  }

  /**
   * Set time bounds
   */
  withTimeBounds(bounds: TimeBounds): PolicyConfig {
    this.timeBounds = bounds;
    return this;
  }

  /**
   * Set business hours (9 AM - 5 PM UTC, weekdays)
   */
  withBusinessHours(): PolicyConfig {
    this.timeBounds = {
      startHour: 9,
      endHour: 17,
      allowedDays: [1, 2, 3, 4, 5],
    };
    return this;
  }

  /**
   * Set contract restrictions
   */
  withContractRestrictions(restrictions: ContractRestriction): PolicyConfig {
    this.contractRestrictions = restrictions;
    return this;
  }

  /**
   * Set additional approval threshold
   */
  withAdditionalApprovalThreshold(amount: bigint): PolicyConfig {
    this.additionalApprovalThreshold = amount;
    return this;
  }

  /**
   * Convert to JSON for storage
   */
  toJSON(): object {
    return {
      spendingLimits: Object.fromEntries(
        Array.from(this.spendingLimits.entries()).map(([k, v]) => [
          k,
          {
            ...v,
            perTransaction: v.perTransaction?.toString(),
            daily: v.daily?.toString(),
            weekly: v.weekly?.toString(),
          },
        ])
      ),
      whitelist: this.whitelist ? Array.from(this.whitelist) : null,
      blacklist: Array.from(this.blacklist),
      timeBounds: this.timeBounds,
      contractRestrictions: this.contractRestrictions
        ? {
            allowedContracts: Array.from(
              this.contractRestrictions.allowedContracts
            ),
            allowedSelectors: Array.from(
              this.contractRestrictions.allowedSelectors
            ),
            blockedSelectors: Array.from(
              this.contractRestrictions.blockedSelectors
            ),
          }
        : null,
      additionalApprovalThreshold:
        this.additionalApprovalThreshold?.toString() ?? null,
      maxPendingRequests: this.maxPendingRequests,
      enabled: this.enabled,
    };
  }

  /**
   * Create from JSON
   */
  static fromJSON(json: object): PolicyConfig {
    const config = new PolicyConfig();
    const data = json as Record<string, unknown>;

    config.enabled = (data.enabled as boolean) ?? true;

    if (data.whitelist && Array.isArray(data.whitelist)) {
      config.whitelist = new Set(data.whitelist as string[]);
    }

    if (data.blacklist && Array.isArray(data.blacklist)) {
      config.blacklist = new Set(data.blacklist as string[]);
    }

    if (data.timeBounds) {
      config.timeBounds = data.timeBounds as TimeBounds;
    }

    if (data.additionalApprovalThreshold) {
      config.additionalApprovalThreshold = BigInt(
        data.additionalApprovalThreshold as string
      );
    }

    return config;
  }
}

/**
 * Policy engine for evaluating transactions
 */
export class PolicyEngine {
  private config: PolicyConfig;
  private dailySpending: Map<string, bigint> = new Map();
  private weeklySpending: Map<string, bigint> = new Map();

  constructor(config: PolicyConfig) {
    this.config = config;
  }

  /**
   * Get the policy configuration
   */
  getConfig(): PolicyConfig {
    return this.config;
  }

  /**
   * Update the policy configuration
   */
  setConfig(config: PolicyConfig): void {
    this.config = config;
  }

  /**
   * Evaluate a transaction against the policy
   */
  evaluate(tx: TransactionRequest): PolicyDecision {
    // Skip evaluation if policy is disabled
    if (!this.config.enabled) {
      return { approved: true, requiresAdditionalApproval: false };
    }

    const toAddress = tx.to.toLowerCase();

    // Check blacklist
    if (this.config.blacklist.has(toAddress)) {
      return {
        approved: false,
        requiresAdditionalApproval: false,
        reason: `Address ${tx.to} is blacklisted`,
      };
    }

    // Check whitelist
    if (this.config.whitelist && !this.config.whitelist.has(toAddress)) {
      return {
        approved: false,
        requiresAdditionalApproval: false,
        reason: `Address ${tx.to} is not whitelisted`,
      };
    }

    // Check time bounds
    if (this.config.timeBounds) {
      const now = new Date();
      const hour = now.getUTCHours();
      const day = now.getUTCDay();

      const { startHour, endHour, allowedDays } = this.config.timeBounds;

      const hourOk =
        startHour <= endHour
          ? hour >= startHour && hour < endHour
          : hour >= startHour || hour < endHour;

      if (!hourOk || !allowedDays.includes(day)) {
        return {
          approved: false,
          requiresAdditionalApproval: false,
          reason: `Transaction outside allowed time window (${startHour}:00-${endHour}:00 UTC)`,
        };
      }
    }

    // Check contract restrictions
    if (tx.data && this.config.contractRestrictions) {
      const restrictions = this.config.contractRestrictions;

      // Check allowed contracts
      if (
        restrictions.allowedContracts.size > 0 &&
        !restrictions.allowedContracts.has(toAddress)
      ) {
        return {
          approved: false,
          requiresAdditionalApproval: false,
          reason: `Contract ${tx.to} is not in allowed list`,
        };
      }

      // Check function selectors
      const selector = tx.data.slice(0, 10).toLowerCase();

      if (restrictions.blockedSelectors.has(selector)) {
        return {
          approved: false,
          requiresAdditionalApproval: false,
          reason: `Function selector ${selector} is blocked`,
        };
      }

      if (
        restrictions.allowedSelectors.size > 0 &&
        !restrictions.allowedSelectors.has(selector)
      ) {
        return {
          approved: false,
          requiresAdditionalApproval: false,
          reason: `Function selector ${selector} is not in allowed list`,
        };
      }
    }

    // Parse transaction value
    const value = this.parseValue(tx.value);

    // Check spending limits
    const limits = this.config.spendingLimits.get(tx.chain);
    if (limits) {
      // Per-transaction limit
      if (limits.perTransaction && value > limits.perTransaction) {
        return {
          approved: false,
          requiresAdditionalApproval: false,
          reason: `Transaction value exceeds per-transaction limit of ${limits.perTransaction}`,
        };
      }

      // Daily limit
      if (limits.daily) {
        const dateKey = new Date().toISOString().split('T')[0];
        const spent = this.dailySpending.get(dateKey) ?? 0n;
        if (spent + value > limits.daily) {
          return {
            approved: false,
            requiresAdditionalApproval: false,
            reason: `Transaction would exceed daily limit of ${limits.daily}`,
          };
        }
      }

      // Weekly limit
      if (limits.weekly) {
        const now = new Date();
        const weekKey = `${now.getFullYear()}-W${Math.ceil(
          now.getDate() / 7
        )}`;
        const spent = this.weeklySpending.get(weekKey) ?? 0n;
        if (spent + value > limits.weekly) {
          return {
            approved: false,
            requiresAdditionalApproval: false,
            reason: `Transaction would exceed weekly limit of ${limits.weekly}`,
          };
        }
      }
    }

    // Check additional approval threshold
    if (
      this.config.additionalApprovalThreshold &&
      value > this.config.additionalApprovalThreshold
    ) {
      return {
        approved: false,
        requiresAdditionalApproval: true,
        reason: `Transaction value exceeds additional approval threshold`,
      };
    }

    return { approved: true, requiresAdditionalApproval: false };
  }

  /**
   * Record a completed transaction for spending tracking
   */
  recordTransaction(tx: TransactionRequest): void {
    const value = this.parseValue(tx.value);
    const dateKey = new Date().toISOString().split('T')[0];
    const now = new Date();
    const weekKey = `${now.getFullYear()}-W${Math.ceil(now.getDate() / 7)}`;

    // Update daily spending
    const dailySpent = this.dailySpending.get(dateKey) ?? 0n;
    this.dailySpending.set(dateKey, dailySpent + value);

    // Update weekly spending
    const weeklySpent = this.weeklySpending.get(weekKey) ?? 0n;
    this.weeklySpending.set(weekKey, weeklySpent + value);
  }

  /**
   * Get current daily spending
   */
  getDailySpending(): bigint {
    const dateKey = new Date().toISOString().split('T')[0];
    return this.dailySpending.get(dateKey) ?? 0n;
  }

  /**
   * Get current weekly spending
   */
  getWeeklySpending(): bigint {
    const now = new Date();
    const weekKey = `${now.getFullYear()}-W${Math.ceil(now.getDate() / 7)}`;
    return this.weeklySpending.get(weekKey) ?? 0n;
  }

  /**
   * Reset spending trackers
   */
  resetSpending(): void {
    this.dailySpending.clear();
    this.weeklySpending.clear();
  }

  /**
   * Parse a value string to bigint
   */
  private parseValue(value: string): bigint {
    // Handle decimal values (e.g., "1.5" ETH -> wei)
    if (value.includes('.')) {
      const [whole, decimal] = value.split('.');
      const wholeBigInt = BigInt(whole || '0') * 10n ** 18n;
      const decimalPadded = decimal.padEnd(18, '0').slice(0, 18);
      const decimalBigInt = BigInt(decimalPadded);
      return wholeBigInt + decimalBigInt;
    }
    return BigInt(value);
  }
}

File: packages/mpc-wallet-python/src/mpc_wallet/policy.py
"""Policy engine for transaction validation."""

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any

from .types import ChainType, TransactionRequest


@dataclass
class PolicyDecision:
    """Policy decision result."""

    approved: bool
    requires_additional_approval: bool = False
    reason: str | None = None

    def can_proceed(self) -> bool:
        """Check if transaction should proceed."""
        return self.approved or self.requires_additional_approval


@dataclass
class SpendingLimits:
    """Spending limits configuration."""

    per_transaction: int | None = None  # In wei
    daily: int | None = None
    weekly: int | None = None
    currency: str = "ETH"


@dataclass
class TimeBounds:
    """Time window restriction."""

    start_hour: int = 0  # 0-23, UTC
    end_hour: int = 24  # 0-23, UTC
    allowed_days: list[int] = field(default_factory=lambda: [0, 1, 2, 3, 4, 5, 6])

    @classmethod
    def business_hours(cls) -> "TimeBounds":
        """Create business hours restriction (9 AM - 5 PM UTC, weekdays)."""
        return cls(start_hour=9, end_hour=17, allowed_days=[1, 2, 3, 4, 5])

    def is_allowed(self, dt: datetime) -> bool:
        """Check if a datetime falls within the time bounds."""
        hour = dt.hour
        day = dt.weekday()  # 0=Monday in Python, need to convert

        # Convert Python weekday (0=Monday) to JS-style (0=Sunday)
        day = (day + 1) % 7

        if self.start_hour <= self.end_hour:
            hour_ok = self.start_hour <= hour < self.end_hour
        else:
            # Handles wrap-around (e.g., 22:00 - 06:00)
            hour_ok = hour >= self.start_hour or hour < self.end_hour

        return hour_ok and day in self.allowed_days


@dataclass
class ContractRestriction:
    """Contract interaction restriction."""

    allowed_contracts: set[str] = field(default_factory=set)
    allowed_selectors: set[str] = field(default_factory=set)
    blocked_selectors: set[str] = field(default_factory=set)

    def allow_contract(self, address: str) -> "ContractRestriction":
        """Add an allowed contract."""
        self.allowed_contracts.add(address.lower())
        return self

    def allow_selector(self, selector: str) -> "ContractRestriction":
        """Add an allowed function selector."""
        self.allowed_selectors.add(selector.lower())
        return self

    def block_selector(self, selector: str) -> "ContractRestriction":
        """Block a function selector."""
        self.blocked_selectors.add(selector.lower())
        return self


@dataclass
class PolicyConfig:
    """Policy configuration."""

    spending_limits: dict[ChainType, SpendingLimits] = field(default_factory=dict)
    whitelist: set[str] | None = None
    blacklist: set[str] = field(default_factory=set)
    time_bounds: TimeBounds | None = None
    contract_restrictions: ContractRestriction | None = None
    additional_approval_threshold: int | None = None
    max_pending_requests: int = 10
    enabled: bool = True

    @classmethod
    def disabled(cls) -> "PolicyConfig":
        """Create a disabled policy (all transactions allowed)."""
        return cls(enabled=False)

    def with_per_tx_limit(self, amount: int, currency: str = "ETH") -> "PolicyConfig":
        """Set per-transaction spending limit."""
        limits = self.spending_limits.get(ChainType.EVM, SpendingLimits())
        limits.per_transaction = amount
        limits.currency = currency
        self.spending_limits[ChainType.EVM] = limits
        return self

    def with_daily_limit(self, amount: int) -> "PolicyConfig":
        """Set daily spending limit."""
        limits = self.spending_limits.get(ChainType.EVM, SpendingLimits())
        limits.daily = amount
        self.spending_limits[ChainType.EVM] = limits
        return self

    def with_weekly_limit(self, amount: int) -> "PolicyConfig":
        """Set weekly spending limit."""
        limits = self.spending_limits.get(ChainType.EVM, SpendingLimits())
        limits.weekly = amount
        self.spending_limits[ChainType.EVM] = limits
        return self

    def with_whitelist(self, addresses: list[str]) -> "PolicyConfig":
        """Set address whitelist."""
        self.whitelist = {a.lower() for a in addresses}
        return self

    def with_blacklist(self, addresses: list[str]) -> "PolicyConfig":
        """Set address blacklist."""
        self.blacklist = {a.lower() for a in addresses}
        return self

    def with_time_bounds(self, bounds: TimeBounds) -> "PolicyConfig":
        """Set time bounds."""
        self.time_bounds = bounds
        return self

    def with_business_hours(self) -> "PolicyConfig":
        """Set business hours restriction."""
        self.time_bounds = TimeBounds.business_hours()
        return self

    def with_contract_restrictions(self, restrictions: ContractRestriction) -> "PolicyConfig":
        """Set contract restrictions."""
        self.contract_restrictions = restrictions
        return self

    def with_additional_approval_threshold(self, amount: int) -> "PolicyConfig":
        """Set additional approval threshold."""
        self.additional_approval_threshold = amount
        return self


class PolicyEngine:
    """Policy engine for evaluating transactions."""

    def __init__(self, config: PolicyConfig) -> None:
        self._config = config
        self._daily_spending: dict[str, int] = {}
        self._weekly_spending: dict[str, int] = {}

    @property
    def config(self) -> PolicyConfig:
        """Get the policy configuration."""
        return self._config

    def set_config(self, config: PolicyConfig) -> None:
        """Update the policy configuration."""
        self._config = config

    def evaluate(self, tx: TransactionRequest) -> PolicyDecision:
        """Evaluate a transaction against the policy."""
        # Skip evaluation if policy is disabled
        if not self._config.enabled:
            return PolicyDecision(approved=True)

        to_address = tx.to.lower()

        # Check blacklist
        if to_address in self._config.blacklist:
            return PolicyDecision(
                approved=False,
                reason=f"Address {tx.to} is blacklisted",
            )

        # Check whitelist
        if self._config.whitelist and to_address not in self._config.whitelist:
            return PolicyDecision(
                approved=False,
                reason=f"Address {tx.to} is not whitelisted",
            )

        # Check time bounds
        if self._config.time_bounds:
            now = datetime.now(timezone.utc)
            if not self._config.time_bounds.is_allowed(now):
                bounds = self._config.time_bounds
                return PolicyDecision(
                    approved=False,
                    reason=f"Transaction outside allowed time window ({bounds.start_hour}:00-{bounds.end_hour}:00 UTC)",
                )

        # Check contract restrictions
        if tx.is_contract_call() and self._config.contract_restrictions:
            restrictions = self._config.contract_restrictions

            # Check allowed contracts
            if restrictions.allowed_contracts and to_address not in restrictions.allowed_contracts:
                return PolicyDecision(
                    approved=False,
                    reason=f"Contract {tx.to} is not in allowed list",
                )

            # Check function selectors
            selector = tx.function_selector()
            if selector:
                selector = selector.lower()

                if selector in restrictions.blocked_selectors:
                    return PolicyDecision(
                        approved=False,
                        reason=f"Function selector {selector} is blocked",
                    )

                if (
                    restrictions.allowed_selectors
                    and selector not in restrictions.allowed_selectors
                ):
                    return PolicyDecision(
                        approved=False,
                        reason=f"Function selector {selector} is not in allowed list",
                    )

        # Parse transaction value
        value = self._parse_value(tx.value)

        # Check spending limits
        limits = self._config.spending_limits.get(tx.chain)
        if limits:
            # Per-transaction limit
            if limits.per_transaction and value > limits.per_transaction:
                return PolicyDecision(
                    approved=False,
                    reason=f"Transaction value exceeds per-transaction limit of {limits.per_transaction}",
                )

            # Daily limit
            if limits.daily:
                now = datetime.now(timezone.utc)
                date_key = now.strftime("%Y-%m-%d")
                spent = self._daily_spending.get(date_key, 0)
                if spent + value > limits.daily:
                    return PolicyDecision(
                        approved=False,
                        reason=f"Transaction would exceed daily limit of {limits.daily}",
                    )

            # Weekly limit
            if limits.weekly:
                now = datetime.now(timezone.utc)
                week_key = now.strftime("%Y-W%W")
                spent = self._weekly_spending.get(week_key, 0)
                if spent + value > limits.weekly:
                    return PolicyDecision(
                        approved=False,
                        reason=f"Transaction would exceed weekly limit of {limits.weekly}",
                    )

        # Check additional approval threshold
        if self._config.additional_approval_threshold:
            if value > self._config.additional_approval_threshold:
                return PolicyDecision(
                    approved=False,
                    requires_additional_approval=True,
                    reason="Transaction value exceeds additional approval threshold",
                )

        return PolicyDecision(approved=True)

    def record_transaction(self, tx: TransactionRequest) -> None:
        """Record a completed transaction for spending tracking."""
        value = self._parse_value(tx.value)
        now = datetime.now(timezone.utc)

        date_key = now.strftime("%Y-%m-%d")
        week_key = now.strftime("%Y-W%W")

        # Update daily spending
        self._daily_spending[date_key] = self._daily_spending.get(date_key, 0) + value

        # Update weekly spending
        self._weekly_spending[week_key] = self._weekly_spending.get(week_key, 0) + value

    def get_daily_spending(self) -> int:
        """Get current daily spending."""
        now = datetime.now(timezone.utc)
        date_key = now.strftime("%Y-%m-%d")
        return self._daily_spending.get(date_key, 0)

    def get_weekly_spending(self) -> int:
        """Get current weekly spending."""
        now = datetime.now(timezone.utc)
        week_key = now.strftime("%Y-W%W")
        return self._weekly_spending.get(week_key, 0)

    def reset_spending(self) -> None:
        """Reset spending trackers."""
        self._daily_spending.clear()
        self._weekly_spending.clear()

    def _parse_value(self, value: str) -> int:
        """Parse a value string to int."""
        # Handle decimal values (e.g., "1.5" ETH -> wei)
        if "." in value:
            whole, decimal = value.split(".")
            whole_int = int(whole or "0") * (10**18)
            decimal_padded = decimal.ljust(18, "0")[:18]
            decimal_int = int(decimal_padded)
            return whole_int + decimal_int
        return int(value)
